File: app/src/main/java/su/linka/pictures/components/GridButton.java
package su.linka.pictures.components;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.StringRes;

import com.amulyakhare.textdrawable.TextDrawable;

import su.linka.pictures.Card;
import su.linka.pictures.R;
import su.linka.pictures.Utils;

public class GridButton  extends LinearLayout {
    private Card card;
    private Bitmap image;
    private boolean isOutputCard;

    public GridButton(Context context){
        this(context, null, null, false);
    }

    public GridButton(Context context, Card card, Bitmap image){
        this(context, card, image, false);
    }
    public GridButton(Context context, Card card, Bitmap image, boolean isOutputCard) {
        super(context);
        this.isOutputCard = isOutputCard;

        inflate(getContext(), R.layout.grid_button, this);
        setCard(card);
        setImage(image);
    }

    public GridButton(Context context, boolean isOutputCard) {
        this(context, null, null, isOutputCard);
    }

    public void setCard(Card card) {
        if(card!=null&&card.cardType!=2){
            setVisibility(View.VISIBLE);

            if(card.cardType==0){
                setText(card.title);
            }
            if(card.cardType==1){
                setText("");
                ((ImageView) findViewById(R.id.image)).setImageDrawable(getContext().getDrawable(R.drawable.ic_baseline_space_bar_24));
            }
            if(card.cardType==3){
                setText(R.string.create_card);
                ((ImageView) findViewById(R.id.image)).setImageDrawable(getContext().getDrawable(R.drawable.ic_baseline_add_24));

            }

        } else {
            setVisibility(View.INVISIBLE);

        }
        this.card = card;
    }

    private void setText(@StringRes int id) {
        setText(getContext().getString(id));
    }

    private void setText(String title) {
        Bitmap bm = Utils.textAsBitmap(title);
        ImageView image = (ImageView) findViewById(R.id.image_text);
        image.setImageBitmap(bm);
    }

    public Card getCard() {

        return card;
    }

    public void setImage(Bitmap image) {
        if(image!=null){

            if(card.cardType==0) {

                ((ImageView) findViewById(R.id.image)).setImageBitmap(image);
            }
        }
        this.image = image;
    }

    public Bitmap getImage() {
        return image;
    }

    @Override
    public void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        if(isOutputCard) {
            super.onMeasure(heightMeasureSpec, heightMeasureSpec);
        } else {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        }
    }
}

File: app/src/main/java/su/linka/pictures/components/GridButton.kt
package su.linka.pictures.components

import android.content.Context
import android.graphics.Bitmap
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import com.bumptech.glide.Glide
import su.linka.pictures.Card
import su.linka.pictures.R
import su.linka.pictures.Utils

class GridButton @JvmOverloads constructor(
    context: Context,
    private var card: Card? = null,
    private var image: Bitmap? = null,
    private val isOutputCard: Boolean = false
) : LinearLayout(context) {

    init {
        inflate(context, R.layout.grid_button, this)
        setCard(card)
        setImage(image)
    }

    fun setCard(card: Card?) {
        this.card = card
        val imageView = findViewById<ImageView>(R.id.image)
        when {
            card == null || card.cardType == 2 -> {
                visibility = View.INVISIBLE
                imageView.setImageDrawable(null)
                setText("")
            }
            card.cardType == 0 -> {
                visibility = View.VISIBLE
                setText(card.title.orEmpty())
            }
            card.cardType == 1 -> {
                visibility = View.VISIBLE
                setText("")
                imageView.setImageDrawable(
                    ContextCompat.getDrawable(context, R.drawable.ic_baseline_space_bar_24)
                )
            }
            card.cardType == 3 -> {
                visibility = View.VISIBLE
                setText(R.string.create_card)
                imageView.setImageDrawable(
                    ContextCompat.getDrawable(context, R.drawable.ic_baseline_add_24)
                )
            }
        }
    }

    private fun setText(@StringRes id: Int) {
        setText(context.getString(id))
    }

    private fun setText(title: String) {
        val bitmap = Utils.textAsBitmap(title)
        val textImage = findViewById<ImageView>(R.id.image_text)
        textImage.setImageBitmap(bitmap)
    }

    fun getCard(): Card? = card

    fun setImage(image: Bitmap?) {
        this.image = image
        val target = findViewById<ImageView>(R.id.image)
        if (card?.cardType == 0) {
            if (image != null) {
                Glide.with(this)
                    .load(image)
                    .into(target)
            } else {
                target.setImageDrawable(null)
            }
        }
    }

    fun getImage(): Bitmap? = image

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        if (isOutputCard) {
            super.onMeasure(heightMeasureSpec, heightMeasureSpec)
        } else {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        }
    }
}
